### PR TITLE
add no-duplicated-branches rule

### LINKED ISSUE #19: Rule: Two branches in a conditional structure should not have exactly the same implementation
RSPEC-1871

### PR DESCRIPTION
**Pull request by stas-vilchik**
fixes #19

### CODE REVIEW

File: docs/rules/no-duplicated-branches.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# no-all-duplicated-branches
```
Comment (vilchik-elena) [inline docs/rules/no-duplicated-branches.md]: @stas-vilchik typo

File: docs/rules/no-duplicated-branches.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  doTheThirdThing();
+}
+else {
+  doTheRest(); 
+}
+```
+
+## Exceptions
+
+Blocks in an `if` chain that contain a single line of code are ignored, as are blocks in a `switc`h statement that
```
Comment (vilchik-elena) [inline docs/rules/no-duplicated-branches.md]: @stas-vilchik typo

File: src/rules/noDuplicatedBranchesRule.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+class Walker extends tslint.RuleWalker {
+  public visitIfStatement(node: ts.IfStatement) {
+    // don't visit `else if` statements
+    if (!node.parent || node.parent.kind !== ts.SyntaxKind.IfStatement) {
+      const branches = this.collectIfBranches(node);
+
+      for (let i = 1; i < branches.length; i++) {
+        if (this.hasRequiredSize(branches[i])) {
+          for (let j = 0; j < i; j++) {
+            if (this.hasRequiredSize(branches[j]) && areEquivalent(branches[i], branches[j])) {
```
Comment (vilchik-elena) [inline src/rules/noDuplicatedBranchesRule.ts]: we can drop this check for `this.hasRequiredSize(branches[j])`

File: src/rules/noDuplicatedBranchesRule.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      }
+    }
+
+    super.visitSwitchStatement(node);
+  }
+
+  private hasRequiredSize(node: ts.Node | ts.Node[]) {
+    let nodes = Array.isArray(node) ? node : node.getChildren();
+
+    // expand case 1: { ... }
```
Comment (vilchik-elena) [inline src/rules/noDuplicatedBranchesRule.ts]: put code in quotes

File: src/rules/noDuplicatedBranchesRule.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    if (nodes.length === 1 && this.isBlock(nodes[0])) {
+      nodes = nodes[0].getChildren();
+    }
+
+    const children = nodes.filter(
```
Comment (vilchik-elena) [inline src/rules/noDuplicatedBranchesRule.ts]: rename to `statements`

File: src/rules/noDuplicatedBranchesRule.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  private isIfStatement(node: ts.Statement): node is ts.IfStatement {
+    return node.kind === ts.SyntaxKind.IfStatement;
+  }
+
+  private isBlock(node: ts.Node): node is ts.Block {
+    return node.kind === ts.SyntaxKind.Block;
+  }
+
+  private getLine(node: ts.Node): number {
+    return this.getSourceFile().getLineAndCharacterOfPosition(node.getStart(this.getSourceFile())).line + 1;
```
Comment (vilchik-elena) [inline src/rules/noDuplicatedBranchesRule.ts]: remove `this.getSourceFile()`

File: src/rules/noDuplicatedBranchesRule.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  public visitIfStatement(node: ts.IfStatement) {
+    // don't visit `else if` statements
+    if (!node.parent || node.parent.kind !== ts.SyntaxKind.IfStatement) {
+      const branches = this.collectIfBranches(node);
+
+      for (let i = 1; i < branches.length; i++) {
+        if (this.hasRequiredSize(branches[i])) {
+          for (let j = 0; j < i; j++) {
+            if (this.hasRequiredSize(branches[j]) && areEquivalent(branches[i], branches[j])) {
+              this.addFailureAtNode(branches[i], Rule.formatMessage("branch", this.getLine(branches[j])));
```
Comment (vilchik-elena) [inline src/rules/noDuplicatedBranchesRule.ts]: we should break inner loop. same for switch

File: src/rules/noDuplicatedBranchesRule.ts [hunk 8]
```diff
@@ -136,9 +133,14 @@ class Walker extends tslint.RuleWalker {
...
 
-  private takeWithoutBreak(nodes: ts.NodeArray<ts.Node>) {
+  private takeWithoutBreak(nodes: ts.Node[]) {
     return nodes.length > 0 && nodes[nodes.length - 1].kind === ts.SyntaxKind.BreakStatement
       ? nodes.slice(0, -1)
       : nodes;
   }
+
+  private expandSingleBlockStatement(nodes: ts.Node[]) {
+    const child = nodes[0];
```
Comment (vilchik-elena) [inline src/rules/noDuplicatedBranchesRule.ts]: why no check on nodes number?

Comment (stas-vilchik) [inline src/rules/noDuplicatedBranchesRule.ts]: no need, `child` will be `undefined` and the next check `this.isBlock(child)` will fail

**Pull request merged**